src/compute/src/sink/copy_to_s3_oneshot.rs

```diff
@@ -58,22 +57,29 @@ where
             }
         };
 
-        // Move all the data to a single worker and consolidate
-        // TODO(#7256): split the data among worker to some number of buckets (either static or configurable
-        // by the user) to split the load across the cluster.
-        let scope = sinked_collection.scope();
-        let active_worker = usize::cast_from(sink_id.hashed()) % scope.peers();
-        let exchange = Exchange::new(move |_| u64::cast_from(active_worker));
-        let error_exchange = Exchange::new(move |_| u64::cast_from(active_worker));
+        // Splitting the data across a known number of buckets to distribute load across the cluster.
+        // Each worker will be handling data belonging to 0 or more buckets. We are doing this so that
+        // we can write files to s3 deterministically across different replicas of different sizes
+        // using the bucket ID. Each worker will split a bucket's data into 1 or more
+        // files based on the user provided `MAX_FILE_SIZE`.
+        // TODO(#7256): make the number of buckets configurable.
+        let bucket_count = 20;
+
         let input = consolidate_pact::<KeyBatcher<_, _, _>, _, _, _, _, _>(
-            &sinked_collection.map(|row| (row, ())),
-            exchange,
+            &sinked_collection.map(move |row| {
+                let bucket = row.hashed() % bucket_count;
+                ((row, bucket), ())
+            }),
+            Exchange::new(move |(((_, bucket), _), _, _)| *bucket),
             "Consolidated COPY TO S3 input",
         );
```

Comment:
We will need the data to be in a deterministic order after the exchange. Is that a correct assumption to make?

Reply:
Is there any way to actually get the data in an order? The splitting into multiple files will not work otherwise.

Reply:
I think that's fine though, right? The property we need is that two replicas with different worker counts that are executing the same `COPY TO S3` operation need to put the same rows in the same CSV files. I don't think the order of the rows in those CSV files matters, at long as the multiset of rows is identical.

Reply:
@benesch But it will be a problem if a worker has to split the data into multiple files to stay under MAX_FILE_SIZE? The workers in different replicas will have different data in the splits then.

Reply:
er...should we just ignore the max_file_size for the time being and set like a large value internally like 5GB?

Reply:
The data will come out of this operator in a deterministic order because here we are defining the pact of consolidation which will sort the data. We don't really care that the exchange pact maintains the order (even though it does). @antiguru does that check out?

Reply:
Yes, but it's not a property we support -- it might work right now, but it's relying on something that's not documented as such. I'm fine with leaving it as-is for the moment if there's a follow-up issue filed: Instead of using `consolidate_pact`, the implementation could hook into the merge batcher (as `consolidate_pact` does under the hood) and observe the correctly-ordered stream of updates, without transitioning through a stream.

Reply:
Ack, ty. Filed https://github.com/MaterializeInc/materialize/issues/25835 to track and will also add a comment in the code to that effect.